<a href="https://colab.research.google.com/github/irfan2543/DataScienceMachineLearning/blob/main/Soal2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

2. Menggunakan dataset superstore! Buatlah 3 kelompok 'segment'
(high, medium, dan low) berdasarkan performance Sales yang
digunakan sebagai acuan tim bisnis untuk mengelompokan
customer superstore.

Instruksi:
Bantulah tim bisnis untuk menentukan kriteria pengelompokan untuk 3 segmentasi customer ini, dan gunakan iterrows atau apply
lambda sebagai teknik pembuatan kolom baru!

Berikan alasan kalian dalam penentuan kriteria pengelompokan customer yang dibuat!

Tentukan jumlah customer pada masing-masing segment di tahun 2017!

Download https://drive.google.com/drive/u/1/folders/1cAfoalJUoLugHUuyCPfKiE_eB-eipZ8s

In [ ]:
import pandas as pd
import numpy as np
import datetime as dt

In [ ]:
#Memanggil Data dari Files google drive dan Melihat Info data
'''
url = 'https://drive.google.com/uc?id={}'.format("1SGWnAVxNAwzgXy4JAzjDhv1ygPWLeOqa")
df_super = pd.read_csv(url)
df_super.info()'''

'\nurl = \'https://drive.google.com/uc?id={}\'.format("1SGWnAVxNAwzgXy4JAzjDhv1ygPWLeOqa")\ndf_super = pd.read_csv(url)\ndf_super.info()'

In [ ]:
#Memanggil Data dari Upload File
from google.colab import files
upload_files = files.upload()

Saving superstore.xls to superstore.xls


In [ ]:
import io
xlsx_file = io.BytesIO(upload_files.get('superstore.xls'))

In [ ]:
df_super = pd.read_excel(xlsx_file, 'Orders')
df_super

Row ID        Order ID Order Date  Ship Date       Ship Mode  \
0          1  CA-2016-152156 2016-11-08 2016-11-11    Second Class   
1          2  CA-2016-152156 2016-11-08 2016-11-11    Second Class   
2          3  CA-2016-138688 2016-06-12 2016-06-16    Second Class   
3          4  US-2015-108966 2015-10-11 2015-10-18  Standard Class   
4          5  US-2015-108966 2015-10-11 2015-10-18  Standard Class   
...      ...             ...        ...        ...             ...   
9989    9990  CA-2014-110422 2014-01-21 2014-01-23    Second Class   
9990    9991  CA-2017-121258 2017-02-26 2017-03-03  Standard Class   
9991    9992  CA-2017-121258 2017-02-26 2017-03-03  Standard Class   
9992    9993  CA-2017-121258 2017-02-26 2017-03-03  Standard Class   
9993    9994  CA-2017-119914 2017-05-04 2017-05-09    Second Class   

     Customer ID     Customer Name    Segment        Country             City  \
0       CG-12520       Claire Gute   Consumer  United States        Henderson   
1       CG-12520       Claire Gute   Consumer  United States        Henderson   
2       DV-13045   Darrin Van Huff  Corporate  United States      Los Angeles   
3       SO-20335    Sean O'Donnell   Consumer  United States  Fort Lauderdale   
4       SO-20335    Sean O'Donnell   Consumer  United States  Fort Lauderdale   
...          ...               ...        ...            ...              ...   
9989    TB-21400  Tom Boeckenhauer   Consumer  United States            Miami   
9990    DB-13060       Dave Brooks   Consumer  United States       Costa Mesa   
9991    DB-13060       Dave Brooks   Consumer  United States       Costa Mesa   
9992    DB-13060       Dave Brooks   Consumer  United States       Costa Mesa   
9993    CC-12220      Chris Cortes   Consumer  United States      Westminster   

      ... Postal Code  Region       Product ID         Category Sub-Category  \
0     ...       42420   South  FUR-BO-10001798        Furniture    Bookcases   
1     ...       42420   South  FUR-CH-10000454        Furniture       Chairs   
2     ...       90036    West  OFF-LA-10000240  Office Supplies       Labels   
3     ...       33311   South  FUR-TA-10000577        Furniture       Tables   
4     ...       33311   South  OFF-ST-10000760  Office Supplies      Storage   
...   ...         ...     ...              ...              ...          ...   
9989  ...       33180   South  FUR-FU-10001889        Furniture  Furnishings   
9990  ...       92627    West  FUR-FU-10000747        Furniture  Furnishings   
9991  ...       92627    West  TEC-PH-10003645       Technology       Phones   
9992  ...       92627    West  OFF-PA-10004041  Office Supplies        Paper   
9993  ...       92683    West  OFF-AP-10002684  Office Supplies   Appliances   

                                           Product Name     Sales  Quantity  \
0                     Bush Somerset Collection Bookcase  261.9600         2   
1     Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400         3   
2     Self-Adhesive Address Labels for Typewriters b...   14.6200         2   
3         Bretford CR4500 Series Slim Rectangular Table  957.5775         5   
4                        Eldon Fold 'N Roll Cart System   22.3680         2   
...                                                 ...       ...       ...   
9989                             Ultra Door Pull Handle   25.2480         3   
9990  Tenex B1-RE Series Chair Mats for Low Pile Car...   91.9600         2   
9991                              Aastra 57i VoIP phone  258.5760         2   
9992  It's Hot Message Books with Stickers, 2 3/4" x 5"   29.6000         4   
9993  Acco 7-Outlet Masterpiece Power Center, Wihtou...  243.1600         2   

      Discount    Profit  
0         0.00   41.9136  
1         0.00  219.5820  
2         0.00    6.8714  
3         0.45 -383.0310  
4         0.20    2.5164  
...        ...       ...  
9989      0.20    4.1028  
9990      0.00   15.6332  
9991      0.20   19.3932  
9992      0.

Untuk pengelompokkan data berdasarkan Kategori Norma Hipotetik yang digunakan.
dengan menggunakan Norma Hipotetik menggunakan rumus sebagai berikut :
Mean = Nilai Tertinggi Persentase + Nilai Terendah Persentase / 2 = 100 + 0 / 2 = 50
Standar Deviasi = Nilai Tertinggi Persentase + Nilai Terendah Persentase / 6 = 100 + 0 / 6 = 17 (Dibulatkan)

Jadi dari hasil diatas didapatkan rumus berikut :

Low = X < 50 - 17 = X < 33%
Medium = 50 - 17 >= X <= 50 + 17 = 33% >= X <= 67%
High = X > 50 + 17 = X > 67%

bisa dilihat pada excel data berjumlah 9994 dengan kita asumsikan bahwa persentase untuk perhitungan adalah 100%.
dari data tersebut dapat kita bagi menjadi 3 kategori dalam frekuensi distribusinya yaitu :
  1. Kategori dengan Segment High dengan Range persentase yaitu > 67%
  2. Kateogri dengan Segment Low dengan Range persentase yaitu < 33%
  3. Kategori dengan Segment Medium dengan Range persentase diantara Low dan High yaitu >=33% dan <=67%

In [ ]:
#Membuat range antara nilai persentase maksimum dan minimum untuk Sales
high_percentile = 0.67
low_percentile = 0.33

In [ ]:
#Memanggil nilai dari quantile yang akan di hitung untuk menentukan median dari dataframe
sales_quantile = df_super['Sales'].quantile([high_percentile, low_percentile])

In [ ]:
#Melihat hasil nilai tertinggi dan terendah untuk dataframe
sales_quantile

0.67    129.9200
0.33     24.8707
Name: Sales, dtype: float64

In [ ]:
'''Membuat Fungsi untuk melakukan pembagian antar segment dengan berdasarkan :
  dari nilai maksimum sales dengan data diatas 0.70 dan didapatkan nilai terkecil adalah 129.9200
  dari nilai minimum sales dengan diatas 0.30 dan didapatkan nilai terbesar adalah 24.8707
'''
def segment1(sales):

    if sales > sales_quantile[high_percentile]:
        return "Segment High"
    elif sales > sales_quantile[low_percentile]:
        return "Segment Medium"
    else:
        return "Segment Low"

In [ ]:
#Melakukan pemanggilan menggunakan apply lambda untuk memasukkan fungsi segment kedalam column baru yaitu column Segment Customer
df_super['Segment Customer'] = df_super['Sales'].apply(lambda x: segment1(x))

In [ ]:
df_super

Row ID        Order ID Order Date  Ship Date       Ship Mode  \
0          1  CA-2016-152156 2016-11-08 2016-11-11    Second Class   
1          2  CA-2016-152156 2016-11-08 2016-11-11    Second Class   
2          3  CA-2016-138688 2016-06-12 2016-06-16    Second Class   
3          4  US-2015-108966 2015-10-11 2015-10-18  Standard Class   
4          5  US-2015-108966 2015-10-11 2015-10-18  Standard Class   
...      ...             ...        ...        ...             ...   
9989    9990  CA-2014-110422 2014-01-21 2014-01-23    Second Class   
9990    9991  CA-2017-121258 2017-02-26 2017-03-03  Standard Class   
9991    9992  CA-2017-121258 2017-02-26 2017-03-03  Standard Class   
9992    9993  CA-2017-121258 2017-02-26 2017-03-03  Standard Class   
9993    9994  CA-2017-119914 2017-05-04 2017-05-09    Second Class   

     Customer ID     Customer Name    Segment        Country             City  \
0       CG-12520       Claire Gute   Consumer  United States        Henderson   
1       CG-12520       Claire Gute   Consumer  United States        Henderson   
2       DV-13045   Darrin Van Huff  Corporate  United States      Los Angeles   
3       SO-20335    Sean O'Donnell   Consumer  United States  Fort Lauderdale   
4       SO-20335    Sean O'Donnell   Consumer  United States  Fort Lauderdale   
...          ...               ...        ...            ...              ...   
9989    TB-21400  Tom Boeckenhauer   Consumer  United States            Miami   
9990    DB-13060       Dave Brooks   Consumer  United States       Costa Mesa   
9991    DB-13060       Dave Brooks   Consumer  United States       Costa Mesa   
9992    DB-13060       Dave Brooks   Consumer  United States       Costa Mesa   
9993    CC-12220      Chris Cortes   Consumer  United States      Westminster   

      ... Region       Product ID         Category Sub-Category  \
0     ...  South  FUR-BO-10001798        Furniture    Bookcases   
1     ...  South  FUR-CH-10000454        Furniture       Chairs   
2     ...   West  OFF-LA-10000240  Office Supplies       Labels   
3     ...  South  FUR-TA-10000577        Furniture       Tables   
4     ...  South  OFF-ST-10000760  Office Supplies      Storage   
...   ...    ...              ...              ...          ...   
9989  ...  South  FUR-FU-10001889        Furniture  Furnishings   
9990  ...   West  FUR-FU-10000747        Furniture  Furnishings   
9991  ...   West  TEC-PH-10003645       Technology       Phones   
9992  ...   West  OFF-PA-10004041  Office Supplies        Paper   
9993  ...   West  OFF-AP-10002684  Office Supplies   Appliances   

                                           Product Name     Sales Quantity  \
0                     Bush Somerset Collection Bookcase  261.9600        2   
1     Hon Deluxe Fabric Upholstered Stacking Chairs,...  731.9400        3   
2     Self-Adhesive Address Labels for Typewriters b...   14.6200        2   
3         Bretford CR4500 Series Slim Rectangular Table  957.5775        5   
4                        Eldon Fold 'N Roll Cart System   22.3680        2   
...                                                 ...       ...      ...   
9989                             Ultra Door Pull Handle   25.2480        3   
9990  Tenex B1-RE Series Chair Mats for Low Pile Car...   91.9600        2   
9991                              Aastra 57i VoIP phone  258.5760        2   
9992  It's Hot Message Books with Stickers, 2 3/4" x 5"   29.6000        4   
9993  Acco 7-Outlet Masterpiece Power Center, Wihtou...  243.1600        2   

      Discount    Profit  Segment Customer  
0         0.00   41.9136      Segment High  
1         0.00  219.5820      Segment High  
2         0.00    6.8714       Segment Low  
3         0.45 -383.0310      Segment High  
4         0.20    2.5164       Segment Low  
...        ...       ...               ...  
9989      0.20    4.1028    Segment Medium  
9990      0.00   15.6332    Segment Medium  
9991      0.20   19.3932      Segment High  


In [ ]:
#Melakukan Order Date hanya tahun 2017
df_super['Order Date'] = pd.to_datetime(df_super['Order Date'])
order_date = (df_super['Order Date'].dt.year == 2017)

In [ ]:
#Memanggil df_super pada order_date
df_super[order_date]

Row ID        Order ID Order Date  Ship Date       Ship Mode  \
12        13  CA-2017-114412 2017-04-15 2017-04-20  Standard Class   
23        24  US-2017-156909 2017-07-16 2017-07-18    Second Class   
34        35  CA-2017-107727 2017-10-19 2017-10-23    Second Class   
41        42  CA-2017-120999 2017-09-10 2017-09-15  Standard Class   
43        44  CA-2017-139619 2017-09-19 2017-09-23  Standard Class   
...      ...             ...        ...        ...             ...   
9988    9989  CA-2017-163629 2017-11-17 2017-11-21  Standard Class   
9990    9991  CA-2017-121258 2017-02-26 2017-03-03  Standard Class   
9991    9992  CA-2017-121258 2017-02-26 2017-03-03  Standard Class   
9992    9993  CA-2017-121258 2017-02-26 2017-03-03  Standard Class   
9993    9994  CA-2017-119914 2017-05-04 2017-05-09    Second Class   

     Customer ID    Customer Name      Segment        Country          City  \
12      AA-10480     Andrew Allen     Consumer  United States       Concord   
23      SF-20065  Sandra Flanagan     Consumer  United States  Philadelphia   
34      MA-17560     Matt Abelman  Home Office  United States       Houston   
41      LC-16930   Linda Cazamias    Corporate  United States    Naperville   
43      ES-14080       Erin Smith    Corporate  United States     Melbourne   
...          ...              ...          ...            ...           ...   
9988    RA-19885     Ruben Ausman    Corporate  United States        Athens   
9990    DB-13060      Dave Brooks     Consumer  United States    Costa Mesa   
9991    DB-13060      Dave Brooks     Consumer  United States    Costa Mesa   
9992    DB-13060      Dave Brooks     Consumer  United States    Costa Mesa   
9993    CC-12220     Chris Cortes     Consumer  United States   Westminster   

      ...   Region       Product ID         Category Sub-Category  \
12    ...    South  OFF-PA-10002365  Office Supplies        Paper   
23    ...     East  FUR-CH-10002774        Furniture       Chairs   
34    ...  Central  OFF-PA-10000249  Office Supplies        Paper   
41    ...  Central  TEC-PH-10004093       Technology       Phones   
43    ...    South  OFF-ST-10003282  Office Supplies      Storage   
...   ...      ...              ...              ...          ...   
9988  ...    South  TEC-PH-10004006       Technology       Phones   
9990  ...     West  FUR-FU-10000747        Furniture  Furnishings   
9991  ...     West  TEC-PH-10003645       Technology       Phones   
9992  ...     West  OFF-PA-10004041  Office Supplies        Paper   
9993  ...     West  OFF-AP-10002684  Office Supplies   Appliances   

                                           Product Name    Sales Quantity  \
12                                           Xerox 1967   15.552        3   
23                   Global Deluxe Stacking Chair, Gray   71.372        2   
34                                    Easy-staple paper   29.472        3   
41                                   Panasonic Kx-TS550  147.168        4   
43    Advantus 10-Drawer Portable Organizer, Chrome ...   95.616        2   
...                                                 ...      ...      ...   
9988                    Panasonic KX - TS880B Telephone  206.100        5   
9990  Tenex B1-RE Series Chair Mats for Low Pile Car...   91.960        2   
9991                              Aastra 57i VoIP phone  258.576        2   
9992  It's Hot Message Books with Stickers, 2 3/4" x 5"   29.600        4   
9993  Acco 7-Outlet Masterpiece Power Center, Wihtou...  243.160        2   

      Discount   Profit  Segment Customer  
12         0.2   5.4432       Segment Low  
23         0.3  -1.0196    Segment Medium  
34         0.2   9.9468    Segment Medium  
41         0.2  16.5564      Segment High  
43         0.2   9.5616    Segment Medium  
...        ...      ...               ...  
9988       0.0  55.6470      Segment High  
9990       0.0  15.6332    Segment Medium  
9991       0.2  19.3932      Segment High  
9992       0.0  13.320

In [ ]:
#Melakukan Perhitungan untuk Jumlah Customer pada tahun 2017
hasil_jumlh_customer = df_super[order_date].groupby('Segment Customer')['Order Date'].apply(lambda x: (x=='2017').count()).reset_index(name='Jumlah 2017')

In [ ]:
hasil_jumlh_customer

Segment Customer  Jumlah 2017
0     Segment High         1074
1      Segment Low         1106
2   Segment Medium         1132